stl/inc/ranges

```diff
@@ -620,6 +620,96 @@ namespace ranges {
         /* [[no_unique_address]] */ _Ty _Value{};
     };
 
+    // CLASS TEMPLATE ranges::_Non_propagating_cache
+    // clang-format off
+    template <class _Ty>
+        requires is_object_v<_Ty>
+    class _Non_propagating_cache { // a simplified optional that augments copy_constructible types with full copyability
+        // clang-format on
+    public:
+        constexpr _Non_propagating_cache() noexcept : _Dummy{} {}
+
+        template <class... _Types>
+        constexpr _Non_propagating_cache(in_place_t, _Types&&... _Args) noexcept(
+            is_nothrow_constructible_v<_Ty, _Types...>) // strengthened
+            : _Val(_STD forward<_Types>(_Args)...), _Engaged{true} {}
+
+        _CONSTEXPR20_DYNALLOC ~_Non_propagating_cache() {
+            if (_Engaged) {
+                _Val.~_Ty();
+            }
+        }
+
+        constexpr _Non_propagating_cache(const _Non_propagating_cache&) noexcept {}
+
+        _CONSTEXPR20_DYNALLOC _Non_propagating_cache(_Non_propagating_cache&& _Other) noexcept {
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Other._Engaged = false;
+            }
+        }
+
+        _CONSTEXPR20_DYNALLOC _Non_propagating_cache& operator=(const _Non_propagating_cache& _Other) noexcept {
+            if (_STD addressof(_Other) == this) {
+                return *this;
+            }
+
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+
+            return *this;
+        }
+
+        _CONSTEXPR20_DYNALLOC _Non_propagating_cache& operator=(_Non_propagating_cache&& _Other) noexcept {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Other._Engaged = false;
+            }
+
+            return *this;
+        }
+
+        constexpr explicit operator bool() const noexcept {
+            return _Engaged;
+        }
+
+        _NODISCARD constexpr _Ty& operator*() noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+        _NODISCARD constexpr const _Ty& operator*() const noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+
+        template <class _It>
+        constexpr _Ty& _Emplace_deref(_It& _Iter) noexcept(noexcept(_Ty(*_Iter))) /* strengthened */ {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+
+            _Construct_in_place(_Val, *_Iter);
+            _Engaged = true;
+
+            return _Val;
+        }
+
+    private:
+        union {
+            _Nontrivial_dummy_type _Dummy;
+            _Ty _Val;
+        };
+        bool _Engaged = false;

```

Comment:
We should have a specialization for trivially destructible types that don't need the `_Engaged` member.

Reply:
I had one initially, but the compilers complained that one cannot have a redeclaration of a template with mismatched requires clause 

I think the solution would be to create an `_Object` concept and use that in the template header so that we only use a requires clause for the trivially_destructible specialization

Reply:
Specialization is not a redeclaration. Are you sure you didn't forget the `<_Ty>` in:
```
template <class _Ty>
        requires is_trivially_destructible_v<_Ty>
class _Non_propagating_cache<_Ty> /* <- this <_Ty> */ {
```

Reply:
No way, that would be totally embarrassing wouldn't it?

Reply:
```
C:\STL\out\build\x64\out\inc\ranges(626,18): note: similar constraint expressions not considered equivalent; constraint expressions cannot be considered equivalent unless they originate from the same concept
        requires is_object_v<_Ty>
                 ^~~~~~~~~~~~~~~~
C:\STL\out\build\x64\out\inc\ranges(715,18): note: similar constraint expression here
        requires is_object_v<_Ty> && is_trivially_destructible_v<_Ty>
```

Reply:
That's weird, this code compiles fine: https://godbolt.org/z/eG684nvzY